File: tests/unit/test_standardizer.py
import numpy as np

from bolr.adaptation.standardizer import EWStandardizer
from bolr.config.foundation import SurpriseStandardizerConfig


def test_ew_standardizer_uses_prior_statistics_causally() -> None:
    standardizer = EWStandardizer(SurpriseStandardizerConfig(decay=0.5, variance_floor=1e-4, warmup_count=0, clip_z=None))
    state = standardizer.initial_state()
    state, first = standardizer.step(2.0, state)
    assert first["mean_before"] == 0.0
    state, second = standardizer.step(4.0, state)
    assert np.isclose(second["mean_before"], 1.0)
    assert second["z_score"] > 0.0


def test_ew_standardizer_missing_value_holds_state() -> None:
    standardizer = EWStandardizer()
    state = standardizer.initial_state()
    next_state, diagnostics = standardizer.step(None, state)
    assert next_state == state
    assert diagnostics["missing"] is True

File: csrc/tests/test_standardizer.c
#include "test_suite.h"

#include "bolr/standardizer.h"

#include <math.h>

int test_standardizer(void) {
    bolr_standardizer_config config = {0.5, 1e-4, 0, 0.0, 0};
    bolr_standardizer_state state;
    bolr_standardizer_diagnostics first;
    bolr_standardizer_diagnostics second;
    bolr_standardizer_state_init(&config, &state);
    if (bolr_standardizer_step(&config, &state, 2.0, 1, &first) != BOLR_OK) return 1;
    if (bolr_standardizer_step(&config, &state, 4.0, 1, &second) != BOLR_OK) return 1;
    if (fabs(first.z_score - 200.0) > 1e-9) return 1;
    if (!(second.z_score > 2.0)) return 1;
    if (bolr_standardizer_step(&config, &state, 0.0, 0, &second) != BOLR_OK) return 1;
    return second.missing ? 0 : 1;
}
